### ISSUE TITLE
Share a JS File object with the Pyodide kernel to avoid loading the file into memory

### ISSUE THREAD

**Issue by juntyr** [jupyterlite/jupyterlite#1088]
I am trying to process a large user-provided file within Python code running in the Pyodide-based kernel.

ipywidgets have the  [`FileUpload`](https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#file-upload) widget that allows the user to upload a file that is then accessible inside Python. However, this widget reads the file into memory first, taking up precious memory space and limiting the size of the files that can be used.

Pyodide provides access to emscripten's [WORKERFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#workerfs), which can be used to mount a [JS File](https://developer.mozilla.org/en-US/docs/Web/API/File) or Blob object into the file system without reading it into memory. Is there a way to build a widget that uses the standard `<input type="file">` element to let the user provide read-only access to file, and to then share the JS File object with the web worker running the Pyodide kernel such that custom code there could mount a WORKERFS file system such that the Python code gets access to the file but it is never loaded into the sparse RAM we have?

TLDR: Can a JS File be shared from the main thread with the web worker running the Pyodide kernel, e.g. using a JupyterLite-specific ipywidget?

**Comment by juntyr** [comment]
I've been playing around with JupyterLite in the dev tools a bit and made some progress, which can be found in this Jupyter notebook gist: https://gist.github.com/juntyr/b78392f990d423bb835e49dd4e23a260

At the moment, I need to add a breakpoint to the JupyterLite init to capture a reference to the web worker that runs the Pyodide kernel. This is the major remaining problem area, where I'd appreciate some pointers on how we could get that in a more plausible manner.

Once, we have the worker, we can do the following:
1. Create a HTML file upload input element
2. Register a message hook in the worker
3. Post the `uploader.files` to the worker and catching the message there
4. Using `pyodide_js` to create a new mount for a WORKERFS that includes the uploaded files.

The fantastic part is that I can upload a file that exceeds my RAM and then read it within Pyodide. How could I automate this process?

**Comment by juntyr** [comment]
I have made some significant progress using the `BroadcastChannel` API and am now able to automatically mount files uploaded via an input element into the running Pyodide kernel FS. I've updated the gist to reflect this change. I think I should now be able to turn this into a widget and share it with the community :)

**Comment by juntyr** [comment]
> Sounds like good stuff: while I don't think it would be appropriate to add this kind of JupyterLite-specific capability to [ipylab](https://github.com/jtpio/ipylab), a `ipylite` widget library (which _could_ live in this org) that built on top of `ipylab` _would_ make a lot of sense. It would be able to built on top of the existing, tested `comm` bridge across the kernel/server/client worker gaps, rather than adding any more unique pipes, messages, workers, etc. than needed.

I like that idea! I've now managed to get my idea working as a custom `ipywidget`, which is published at https://github.com/juntyr/ipyfilite / https://pypi.org/project/ipyfilite / https://www.npmjs.com/package/ipyfilite. If a larger `ipylitewidget` library is developed, I'd gladly contribute this widget back to it.

I'm wondering, does the `comm` bridge support passing around JS objects without serialising them, at least within JupyterLite? At the moment the zero-copy aspect of my widget depends on the `FileList` being transferred to the Pyodide worker without loading the underlying files into RAM.

**Comment by bollwyvl** [comment]
Thanks for sharing. A few things there
- adding new things to the `js` object from upstream is of course _possible_, but probably not to be _recommended_
- `@lumino/coreutils` ships a `uuidv4` implementation, used in core
  - but there may be existing ids, like the `comm_id` which would be more appropriate than creating _new_ `uuids`. 
- the approach probably generates a _lot_ of `new BroadcastChannel` instances, with no particular way to clean them up
  - i don't know the practical implications of this

>>  any more unique pipes, messages, workers, etc. **than needed.**
>
> support passing around JS objects

Right: `comm` doesn't, and it's clear _some_ more pipes _are_ needed, but perhaps only one client `BroadcastChannel` per kernel launched, for example, which could be shared by multiple widgets.

More broadly, much like `ipylab`'s `JupyterFrontEnd` (which in turn has a `CommandRegistry`), having a `LiteServer` which provided some capability enumeration would help catch the always-going-to-fail case when features like `BroadcastChannel`... or even reaching _well_ outside the box, and making requests to the lite server's REST API, would be possible.

Long story, of course, is getting more lite-compatible kernels working with these kinds of tools... we have a [naive JS implementation](https://github.com/jupyterlite/jupyterlite/pull/141), but seeing this work in other kernels that worked with the same client pieces would be really smashing.

**Comment by juntyr** [comment]
> * adding new things to the `js` object from upstream is of course _possible_, but probably not to be _recommended_

Good point, I can see two alternative places to store it

* store the private global inside the `ipyfilite` package, meaning we might accidentally recreate it if the package is reloaded
* store it somewhere with the Pyodide kernel's comms since that should live throughout the kernel's lifetime, e.g. as a new comm in `pyodide_kernel.kernel_instance.comm_manager`?

What place would you recommend?

> * `@lumino/coreutils` ships a `uuidv4` implementation, used in core

Thanks for the recommendation!

> * but there may be existing ids, like the `comm_id` which would be more appropriate than creating _new_ `uuids`.

Early on in experimentation I was trying to get something akin to the kernel's UUID and just read that in on both the client JS and backend Python side. Could you perhaps give me some pointers on where I would read those from? At the moment I'm generating a new UUID on the Python side on package load and then syncing that to the JS widget, which is a bit clunky.

> * the approach probably generates a _lot_ of `new BroadcastChannel` instances, with no particular way to clean them up

The one on the Python side is already cached to only have one, but you are right that I'm creating a new one for each upload on the JS side at the moment. Could you suggest a nice global place to store it (perhaps again close to existing comms)?

Thank you for your feedback!

If you do add a channel upon kernel creation, it would not even need to be a broadcast channel. I just used that one because it is globally constructible so I wouldn't have to hack around even more. However, just a regular bidirectional channel that's private to each kernel but whose endpoints are exposed e.g. through the existing comms mechanism on both sides would probably be better.

**Comment by bollwyvl** [comment]
> store it somewhere with the Pyodide kernel

There is an API for doing this on any `HasTraits`, of which both the `Interpreter` and `Kernel` are subclasses. Putting shared functionality inside a "manager" class has some nice advantages, so, for example, in e.g. `manager.py`:

```py
# manager.py
from traitlets.config import LoggingConfigurable


class LiteAppManager(LoggingConfigurable):
    pass  # do some stuff here
```

And then in the run-once code:

```py
from traitlets import Instance
from IPython import get_ipython

from .manager import LiteAppManager


def _install_manager_exactly_once():
    ip = get_ipython()
    if hasattr(ip, "lite_app_manager"):
        return
    manager = LiteAppManager(parent=ip) # setting `parent` has some nice benefits
    ip.add_traits(
        lite_app_manager=Instance(LiteAppManager, default_value=LiteAppManager)
    )
```

This will persist for the duration of the kernel lifecycle (which in the lite implementation is _also_ the lifecycle of the worker and emscripten vm), which should be "permanent enough," and `self.parent` would provide access to any of the `InteractiveShell` pieces, including `.kernel`, and any critical IDs needed there.

> syncing that to the JS widget, which is a bit clunky.

I wouldn't worry about _clunky_, if it's _correct_: having a `channel_name` trait (derived from really _anything_, e.g. `id(kernel)`) and requiring the client to provide the information it needs is hardly the worst thing in the world.  By design, kernels generally _don't_ know about their sessions. With the `@ipywidgets.register` decorator, it's possible to create new widgets on the client, but it would again be fine to wait for the kernel to reply back that it has actually created a pipe on which to listen by setting the `channel_name`.

> nice global place to store it (perhaps again close to existing comms)?

A typical pattern is to have some "dirty" state inside a `Private` namespace, e.g.

```ts
# widgets.ts
namespace Private {
  const _channels = new Map<string, BroadcastChannel>();

  export function getBroadcastChannel(key: string): BroadcastChannel {
    let channel = _channels.get(key);
    if (channel == null) {
      channel = new BroadcastChannel(key);
      _channels.set(key, channel);
    }
    return channel;
  }

  export function removeBroadcastChannel(key: string): void {
    _channels.delete(key);
  }
}
```

Clean up will be... harder, probably need to do some more thinking, e.g. add a typed proxy the adding/removing of handlers, maintaining _another_ map of `handlers`, which, once emptied, would let be enough information to destroy the channel, such that the `BroadcastChannel` itself never "escapes" the `Private` boundary.
